### PR TITLE
Columns of PinnedTags can have Include additional tags for this colum…

### PR DESCRIPTION
**Pull request by nberlee**
…n enabled

This adds not only filter capabilities, which are not relevent to this project like `not` and `all` but also has the `any` field, which is an array where you may add more tags.

Unfortunatly, I was unable to create a precise select query for this, as the database field `data` is not jsonb. I think this is really the most optimal way under these circumstances.

image

### CODE REVIEW

File: src/MastodonConnectionHelper.cs [hunk 1]
```diff
@@ -36,11 +37,31 @@ public static async Task<List<string>> GetPinnedTagsAsync()
...
 
+                JsonElement anyArray = doc.RootElement.GetProperty("any");
+                foreach (JsonElement item in anyArray.EnumerateArray())
+                {
+                    JsonElement value = item.GetProperty("value");
+                    if (value.ValueKind != JsonValueKind.Null)
+                    {
+                        String? valuestring = value.GetString();
+                        if (! valuestring.IsNullOrEmpty())
+                            res.Add(valuestring);
```
Comment (jdhartley) [inline src/MastodonConnectionHelper.cs:59]: Consider checking for duplicates tags here, now that we can't rely on `SELECT DISTINCT` to do that for us

Comment (jdhartley) [reply src/MastodonConnectionHelper.cs:59]: Ope, I commented too soon. I see now there is a `.Distinct()` after this function is executed. I am dismissing this commented!!

As an aside, thanks for these PRs, I am super excited for this functionality and how quickly it went from Issue to PR — cheers!!!

File: src/MastodonConnectionHelper.cs [hunk 2]
```diff
@@ -36,11 +37,31 @@ public static async Task<List<string>> GetPinnedTagsAsync()
         await conn.OpenAsync();
 
         var res = new List<string>();
-        await using var cmd = new NpgsqlCommand("SELECT DISTINCT col->'params'->>'id' FROM web_settings, json_array_elements(data->'columns') col WHERE col->>'id' = 'HASHTAG' AND col->'params'->>'id' IS NOT NULL ORDER BY col->'params'->>'id' ASC", conn);
+        // Column 0: the 'original' tag with was pinned
+        // Column 1: Config of 'Include additional tags for this column' this includes the the tags in 'any' array.
+        await using var cmd = new NpgsqlCommand("SELECT DISTINCT col->'params'->>'id', col->'params'->>'tags' FROM web_settings, json_array_elements(data->'columns') col WHERE col->>'id' = 'HASHTAG' AND col->'params'->>'id' IS NOT NULL ORDER BY col->'params'->>'id' ASC", conn);
```
Comment (g3rv4) [inline src/MastodonConnectionHelper.cs:42]: let's indent this a bit to  make it easier to read

```suggestion
        await using var cmd = new NpgsqlCommand(@"
SELECT DISTINCT col->'params'->>'id', col->'params'->>'tags'
FROM   web_settings, json_array_elements(data->'columns') col
WHERE  col->>'id' = 'HASHTAG'
AND    col->'params'->>'id' IS NOT NULL
ORDER BY col->'params'->>'id' ASC", conn);
```

File: src/MastodonConnectionHelper.cs [hunk 3]
```diff
@@ -36,11 +37,31 @@ public static async Task<List<string>> GetPinnedTagsAsync()
...
+                JsonDocument doc = JsonDocument.Parse(reader.GetString(1));
 
+                JsonElement anyArray = doc.RootElement.GetProperty("any");
+                foreach (JsonElement item in anyArray.EnumerateArray())
+                {
+                    JsonElement value = item.GetProperty("value");
+                    if (value.ValueKind != JsonValueKind.Null)
+                    {
+                        String? valuestring = value.GetString();
+                        if (! valuestring.IsNullOrEmpty())
```
Comment (g3rv4) [inline src/MastodonConnectionHelper.cs:58]: use `var` and  `HasValue()`

```suggestion
                var doc = JsonDocument.Parse(reader.GetString(1));

                var anyArray = doc.RootElement.GetProperty("any");
                foreach (var item in anyArray.EnumerateArray())
                {
                    var value = item.GetProperty("value");
                    if (value.ValueKind != JsonValueKind.Null)
                    {
                        var valuestring = value.GetString();
                        if (valuestring.HasValue())
```

**Review by g3rv4 — COMMENTED** [review]
Works great! just have a couple style requirements

File: src/MastodonConnectionHelper.cs [hunk 4]
```diff
@@ -36,11 +37,31 @@ public static async Task<List<string>> GetPinnedTagsAsync()
...
+                            res.Add(valuestring);
+                    }
+                }
+            }
```
Comment (g3rv4) [inline src/MastodonConnectionHelper.cs:62]:
```suggestion
            if (reader.IsDBNull(1)) continue;
            var doc = JsonDocument.Parse(reader.GetString(1));

            var anyArray = doc.RootElement.GetProperty("any");
            foreach (var item in anyArray.EnumerateArray())
            {
                var value = item.GetProperty("value");
                if (value.ValueKind != JsonValueKind.Null)
                {
                    var valuestring = value.GetString();
                    if (valuestring.HasValue())
                        res.Add(valuestring);
                }
            }
```

Comment (nberlee) [reply src/MastodonConnectionHelper.cs]: Excellent!

File: src/MastodonConnectionHelper.cs (hunk 2 above)
Comment (nberlee) [reply src/MastodonConnectionHelper.cs]: fixed

**Review by g3rv4 — APPROVED** [review]

**Comment by g3rv4** [comment]
released on [v1.0.46](https://github.com/g3rv4/GetMoarFediverse/releases/tag/v1.0.46)

**Pull request merged**